libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h

```diff
@@ -0,0 +1,172 @@
+#  ifndef _LIBCUDACXX_NO_EXCEPTIONS
+        throw ::cuda::cuda_error{__status, "Failed to allocate memory with cudaMalloc."};
+#  else
+        _LIBCUDACXX_UNREACHABLE();
+#  endif
+    }
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p __ptr.
+   * @param __ptr Pointer to be deallocated. Must have been allocated through a call to `allocate`
+   * @param __bytes The number of bytes that was passed to the `allocate` call that returned \p __ptr.
+   * @param __alignment The alignment that was passed to the `allocate` call that returned \p __ptr.
+   */
+  void deallocate(void* __ptr, const size_t __bytes, const size_t __alignment = __default_cuda_malloc_alignment) const
+  {
+    // We need to ensure that the provided alignment matches the minimal provided alignment
+    _LIBCUDACXX_ASSERT(
+      __default_cuda_malloc_alignment <= __alignment && (__alignment % __default_cuda_malloc_alignment == 0),
+      "Invalid alignment passed to cuda_memory_resource::deallocate.");
+    const ::cudaError_t __status = ::cudaFree(__ptr);
+    (void) __status;
+    _LIBCUDACXX_ASSERT(__status == cudaSuccess, "cuda_memory_resource::deallocate failed");
+  }
+
+  /**
+   * @brief Equality comparison with another cuda_memory_resource
+   * @return true
+   */
+  _LIBCUDACXX_NODISCARD_ATTRIBUTE constexpr bool operator==(cuda_memory_resource const&) const noexcept
+  {
+    return true;
+  }
+#  if _CCCL_STD_VER <= 2017
+  /**
+   * @brief Inequality comparison with another cuda_memory_resource
+   * @return false
+   */
+  _LIBCUDACXX_NODISCARD_ATTRIBUTE constexpr bool operator!=(cuda_memory_resource const&) const noexcept
+  {
+    return false;
+  }
+#  endif // _CCCL_STD_VER <= 2017
+
+  /**
+   * @brief Equality comparison between a cuda_memory_resource and another resource
+   *
+   * We want to ensure that we can compare any cuda_memory_resource too a resource_ref that holds a cuda_memory_resource
+   * Therefore, we convert both to a resource_ref<> which is the lowest common denominator. The resource_ref<>
+   * internally checks whether both resources are of the same kind and if so returns the comparison of the two.
+   * Otherwise, the comparison will always return false, which is what we want for any other memory resource.
+   *
+   * @param __lhs The cuda_memory_resource
+   * @param __rhs The resource to compare to
+   * @return Comparison of both resources converted to a resource_ref<>
+   */
+  template <class _Resource>
+  _LIBCUDACXX_NODISCARD_FRIEND auto operator==(cuda_memory_resource const& __lhs, _Resource const& __rhs) noexcept
+    _LIBCUDACXX_TRAILING_REQUIRES(bool)(__different_resource<cuda_memory_resource, _Resource>)
+  {
+    return resource_ref<>{const_cast<cuda_memory_resource&>(__lhs)} == resource_ref<>{const_cast<_Resource&>(__rhs)};
+  }
+#  if _CCCL_STD_VER <= 2017
+  /**
+   * @copydoc cuda_memory_resource::operator==(cuda_memory_resource const&, _Resource const&)
```

Comment:
Question: did you verify that this works ? I recall copydoc requiring the exact prototype (e.g. template params may be necessary)

Reply:
Yeah, I checked again and it seems I need to add `<_Resource>`